src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.Windows.cs

```diff
@@ -41,58 +46,112 @@ private static FileStreamStrategy ChooseStrategyCore(SafeFileHandle handle, File
-            // named pipe server can't impersonate a high privileged client security context
-            // (note that this is the effective default on CreateFile2)
-            flagsAndAttributes |= (Interop.Kernel32.SecurityOptions.SECURITY_SQOS_PRESENT | Interop.Kernel32.SecurityOptions.SECURITY_ANONYMOUS);
+                // Our Inheritable bit was stolen from Windows, but should be set in
+                // the security attributes class.  Don't leave this bit set.
+                share &= ~FileShare.Inheritable;
 
-            using (DisableMediaInsertionPrompt.Create())
-            {
-                Debug.Assert(path != null);
-                return ValidateFileHandle(
+                // Must use a valid Win32 constant here...
+                if (mode == FileMode.Append)
+                    mode = FileMode.OpenOrCreate;
+
+                int flagsAndAttributes = (int)options;
+
+                // For mitigating local elevation of privilege attack through named pipes
+                // make sure we always call CreateFile with SECURITY_ANONYMOUS so that the
+                // named pipe server can't impersonate a high privileged client security context
+                // (note that this is the effective default on CreateFile2)
+                flagsAndAttributes |= (Interop.Kernel32.SecurityOptions.SECURITY_SQOS_PRESENT | Interop.Kernel32.SecurityOptions.SECURITY_ANONYMOUS);
+
+                SafeFileHandle safeFileHandle = ValidateFileHandle(
                     Interop.Kernel32.CreateFile(path, fAccess, share, &secAttrs, mode, flagsAndAttributes, IntPtr.Zero),
                     path,
                     (options & FileOptions.Asynchronous) != 0);
+
+                return safeFileHandle;
+            }
+        }
+
+        private static IntPtr NtCreateFile(string fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, long allocationSize)
+        {
+            uint ntStatus;
+            IntPtr fileHandle;
+
+            const string mandatoryNtPrefix = @"\??\";
+            if (fullPath.StartsWith(mandatoryNtPrefix, StringComparison.Ordinal))
+            {
+                (ntStatus, fileHandle) = Interop.NtDll.CreateFile(fullPath, mode, access, share, options, allocationSize);
+            }
+            else
+            {
+                var vsb = new ValueStringBuilder(stackalloc char[1024]);
+                vsb.Append(mandatoryNtPrefix);
+
+                if (fullPath.StartsWith(@"\\?\", StringComparison.Ordinal)) // NtCreateFile does not support "\\?\" prefix, only "\??\"
+                {
+                    vsb.Append(fullPath.AsSpan(4));
+                }
+                else
+                {
+                    vsb.Append(fullPath);
+                }
+
+                (ntStatus, fileHandle) = Interop.NtDll.CreateFile(vsb.AsSpan(), mode, access, share, options, allocationSize);
+                vsb.Dispose();
+            }
```

Comment:
While 1024 is substancially larger than MAX_PATH, and therefore this code can handle any input that does not begin with "\\?\" correctly (unless the limit is disabled and the app manifested for longer paths, which most of .NET core is not set up to support), for the paths that do begin "\\?\", they are allowed to be much longer than 1024. 

This may or may not be something worth addressing, but I thought it was worth a note.

Reply:
@KevinCathcart 1024 is the initial buffer size, when it's not enough the `ValueStringBuilder.Append` method is going to increase the size accordingly (but not using the stack-allocated buffer which has a limited size, but managed buffer from `ArrayPool`)

https://github.com/dotnet/runtime/blob/119785100b1ce36cffcf0caf2cc3e23cce32b222/src/libraries/Common/src/System/Text/ValueStringBuilder.cs#L186-L211

https://github.com/dotnet/runtime/blob/119785100b1ce36cffcf0caf2cc3e23cce32b222/src/libraries/Common/src/System/Text/ValueStringBuilder.cs#L294-L302